### PR TITLE
WIP: added analysis metadata extraction for .7z and .zip archives

### PR DESCRIPTION
**Pull request by cpritcha**
Refactored metadata. Added extractors.
---------------------------------------------------

- updated metadata tests

Analysis metadata extraction functionality
-------------------------------------------------------

- added analysis extraction tests
- added shapefile import errors
- improved data type guessing for tabular data
- removed link finding functionality (did not work generally)

TODO
=====

- insure analysis meets a prescribed structure

### CODE REVIEW

File: miracle/core/metadata.py [hunk 1]
```diff
@@ -3,80 +3,134 @@
 """
 import csv
 import os
+import collections
+import os.path as p
```
Comment (alee) [inline miracle/core/metadata.py]: Would prefer `from os import path` or just `os.path` in the code, thoughts?

File: miracle/core/extractors.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
+import os
+import pyunpack as p
```
Comment (alee) [inline miracle/core/extractors.py]: would prefer explicit package imports so we don't have to remember the shorthand they were aliased as, same with `import collections as c`

File: miracle/core/tests/test_extractors.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+from .common import BaseMiracleTest
+from ..extractors import Extractor
+from ...settings import MIRACLE_ANALYSIS_DIRECTORY
+
+class ExtractorsTest(BaseMiracleTest):
+    ANALYSIS_PATH = os.path.join(settings.BASE_DIR, 'miracle', 'core', 'tests', 'analyses')
+
+    def setUp(self):
+        files = os.listdir(MIRACLE_ANALYSIS_DIRECTORY)
+        for file in files:
```
Comment (alee) [inline miracle/core/tests/test_extractors.py]: this is shadowing a builtin

File: miracle/core/metadata.py (hunk 1 above)
Comment (cpritcha) [inline miracle/core/metadata.py]: Sounds good. I'll change the imports.

File: miracle/core/extractors.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            other_files = OtherFileGlob.from_files(files, log)
+
+            all_files = other_files + shp_files
+            for file in all_files:
+                fullpath = path.join(root, file)
+                metadata.append(load_metadata(fullpath))
+
+        return AnalysisMetadata(metadata, log)
+
+class Extractor(object):
```
Comment (alee) [inline miracle/core/extractors.py]: As the entry point into metadata extraction, clear docs would be nice on expected usage and API (e.g., do we create one Extractor per dataset upload or should we always use the static `from_archive` factory method to construct Extractors?).

**Comment by alee** [comment]
Looks good overall! Some docs on the main collaborating classes and how they work together would be great if you could add those in. When someone uploads a file to our server is this a rough workflow for what we should do?

1. Analyze the file to figure out what it is (could be a zipfile, or just a single dataset / script)
2. Based on file analysis, either use the Extractor to pull apart a zipfile, and then Analyze all of its contents 
3. Once every file has been Analyzed, we have some logic based on its AnalysisMetadata to put the dataset / script in the right place based on its parent Project and store some of our extracted metadata in the RDBMS.

Note that we should probably introduce [Celery and queueing](http://celery.readthedocs.org/en/latest/django/first-steps-with-django.html) and perform all of this asynchronously so the user isn't sitting and spinning on a webpage.

**Pull request merged**